README.md
horde_rpc
===========

horde_rpc is a client library for accessing the Horde XML-RPC interface from Ruby.

Features
--------

* Retrieve Client records
* Record time for Clients

Requirements
------------

* A working Horde installation

Install
-------

    gem install horde_rpc

Examples
--------

```ruby
  require 'horde_rpc'
  horde = HordeRPC.new 'http://horde.bar.com'
  client = horde.first_client_for_company 'UberClient'
  puts "Client by name 'UberClient' has ID #{client['id']}"

  horde.record_time :client       => client['id'],
                    :date         => Date.today,
                    :hours        => 3.5,
                    :employee     => 'foo@bar.com',
                    :description  => 'Did stuff'
                    :type         => 1
```

Author
------

Original author: Ben Langfeld

Links
-----
* [Source](https://github.com/mojolingo/horde_rpc)
* [Documentation](http://rdoc.info/github/mojolingo/horde_rpc/master/frames)
* [Bug Tracker](https://github.com/mojolingo/horde_rpc/issues)

Note on Patches/Pull Requests
-----------------------------

* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  * If you want to have your own version, that is fine but bump version in a commit by itself so I can ignore when I pull
* Send me a pull request. Bonus points for topic branches.

Copyright
---------

Copyright (c) 2012 Mojo Lingo LLC. MIT licence (see LICENSE for details).


CHANGELOG.md
# develop

# v0.2.0
  * Rename to HordeRPC

# v0.1.0
  * First release!


spec/horde_spec.rb
require 'spec_helper'

describe Horde do
  describe "sending requests" do
    before do
      FakeWeb.clean_registry
    end

    it "should hit the specified domain/port" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo'
    end

    it "should be formatted as XMLRPC" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo'

      FakeWeb.last_request.body.should match(/<methodCall>/)
    end

    it "should invoke the correct method, with the correct parameters, including user/password options" do
      horde = Horde.new 'http://horde.bar.com/rpc/', 'usera', 'passwordb'

      FakeWeb.register_uri :post, "http://usera:passwordb@horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo', :foo => 'bar'

      FakeWeb.last_request.body.should be == "<?xml version=\"1.0\" ?><methodCall><methodName>foo</methodName><params><param><value><struct><member><name>foo</name><value><string>bar</string></value></member></struct></value></param></params></methodCall>\n"
    end

    it "should be able to make a request without options" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo'

      FakeWeb.last_request.body.should be == "<?xml version=\"1.0\" ?><methodCall><methodName>foo</methodName><params/></methodCall>\n"
    end

    it "should be able to make a request with multiple options" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo', 'bar', 'baz'

      FakeWeb.last_request.body.should be == "<?xml version=\"1.0\" ?><methodCall><methodName>foo</methodName><params><param><value><string>bar</string></value></param><param><value><string>baz</string></value></param></params></methodCall>\n"
    end

    it "should return the returned data structure" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request('foo').should be == 'bar'
    end
  end

  describe "fetching a client by ID" do
    it "should return the fetched client data" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      id = "uid=foo@bar.com"

      response = {
        "__key"       => "uid=foo@bar.com",
        "__uid"       => "foo@bar.com",
        "firstname"   => "Joe",
        "lastname"    => "Bloggs",
        "name"        => "Joe Bloggs",
        "email"       => "foo@bar.com",
        "homePhone"   => "404 475 4840",
        "workPhone"   => "404 475 4840",
        "cellPhone"   => "404 475 4840",
        "homeAddress" => "Foo Lane, NY",
        "company"     => "Acme",
        "id"          => "uid=foo@bar.com",
        "__type"      => "Object",
        "source"      => "bar_clients"
      }

      horde.expects(:request).once.with('clients.getClient', id).returns response

      client = horde.get_client_by_id id
      client.should be response
    end
  end

  describe "searching for clients" do
    it "should return a set of client data" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      response = {'Acme' => [
        {
          "__key"       => "uid=foo@bar.com",
          "__uid"       => "foo@bar.com",
          "firstname"   => "Joe",
          "lastname"    => "Bloggs",
          "name"        => "Joe Bloggs",
          "email"       => "foo@bar.com",
          "homePhone"   => "404 475 4840",
          "workPhone"   => "404 475 4840",
          "cellPhone"   => "404 475 4840",
          "homeAddress" => "Foo Lane, NY",
          "company"     => "Acme",
          "id"          => "uid=foo@bar.com",
          "__type"      => "Object",
          "source"      => "bar_clients"
        }]
      }

      horde.expects(:request).once.with('clients.searchClients', ['Acme'], ['company']).returns response

      client = horde.search_clients :company => 'Acme'
      client.should be response
    end
  end

  describe "finding the first client by company name" do
    it "should return the relevant client data" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      client_data = {
        "__key"       => "uid=foo@bar.com",
        "__uid"       => "foo@bar.com",
        "firstname"   => "Joe",
        "lastname"    => "Bloggs",
        "name"        => "Joe Bloggs",
        "email"       => "foo@bar.com",
        "homePhone"   => "404 475 4840",
        "workPhone"   => "404 475 4840",
        "cellPhone"   => "404 475 4840",
        "homeAddress" => "Foo Lane, NY",
        "company"     => "Acme",
        "id"          => "uid=foo@bar.com",
        "__type"      => "Object",
        "source"      => "bar_clients"
      }
      response = {'Acme' => [client_data]}

      horde.expects(:request).once.with('clients.searchClients', ['Acme'], ['company']).returns response

      client = horde.first_client_for_company 'Acme'
      client.should be == client_data
    end

    context "if no matches are found" do
      it "should return nil" do
        horde = Horde.new 'http://horde.bar.com/rpc/'

        response = {'Acme' => []}

        horde.expects(:request).once.with('clients.searchClients', ['Acme'], ['company']).returns response

        client = horde.first_client_for_company 'Acme'
        client.should be nil
      end
    end
  end

  describe "recording time" do
    it "should record time given the requested parameters" do
      horde = Horde.new 'http://horde.bar.com/rpc/'

      options = {
        :date         => Date.today,
        :client       => 'bob@acme.com',
        :type         => 1,
        :hours        => 1.2,
        :description  => 'Did stuff',
        :employee     => 'foo@bar.com'
      }

      horde.expects(:request).once.with('time.recordTime', options)

      horde.record_time options
    end

    %w{date client type hours description employee}.each do |required_option|
      context "without specifying #{required_option}" do
        it "should raise ArgumentError" do
          horde = Horde.new 'http://horde.bar.com/rpc/'

          options = {
            :date         => Date.today,
            :client       => 'bob@acme.com',
            :type         => 1,
            :hours        => 1.2,
            :description  => 'Did stuff',
            :employee     => 'foo@bar.com'
          }
          options.merge! required_option.to_sym => nil

          horde.expects(:request).never

          lambda { horde.record_time options }.should raise_error(ArgumentError)
        end
      end
    end
  end
end


spec/spec_helper.rb
require 'horde_rpc'
require 'mocha'
require 'fakeweb'

Dir[File.dirname(__FILE__) + "/support/**/*.rb"].each {|f| require f}

RSpec.configure do |config|
  config.mock_with :mocha
  config.filter_run :focus => true
  config.run_all_when_everything_filtered = true
end

FakeWeb.allow_net_connect = false


spec/horde_rpc_spec.rb
require 'spec_helper'

describe HordeRPC do
  describe "sending requests" do
    before do
      FakeWeb.clean_registry
    end

    it "should hit the specified domain/port" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo'
    end

    it "should be formatted as XMLRPC" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo'

      FakeWeb.last_request.body.should match(/<methodCall>/)
    end

    it "should invoke the correct method, with the correct parameters, including user/password options" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/', 'usera', 'passwordb'

      FakeWeb.register_uri :post, "http://usera:passwordb@horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo', :foo => 'bar'

      FakeWeb.last_request.body.should be == "<?xml version=\"1.0\" ?><methodCall><methodName>foo</methodName><params><param><value><struct><member><name>foo</name><value><string>bar</string></value></member></struct></value></param></params></methodCall>\n"
    end

    it "should be able to make a request without options" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo'

      FakeWeb.last_request.body.should be == "<?xml version=\"1.0\" ?><methodCall><methodName>foo</methodName><params/></methodCall>\n"
    end

    it "should be able to make a request with multiple options" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request 'foo', 'bar', 'baz'

      FakeWeb.last_request.body.should be == "<?xml version=\"1.0\" ?><methodCall><methodName>foo</methodName><params><param><value><string>bar</string></value></param><param><value><string>baz</string></value></param></params></methodCall>\n"
    end

    it "should return the returned data structure" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      FakeWeb.register_uri :post, "http://horde.bar.com/rpc/",
                           :status => ["200", "OK"],
                           :content_type => "text/xml",
                           :body => <<-RESPONSE
<?xml version="1.0"?>
<methodResponse>
  <params>
    <param>
      <value><string>bar</string></value>
    </param>
  </params>
</methodResponse>
RESPONSE

      horde.request('foo').should be == 'bar'
    end
  end

  describe "fetching a client by ID" do
    it "should return the fetched client data" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      id = "uid=foo@bar.com"

      response = {
        "__key"       => "uid=foo@bar.com",
        "__uid"       => "foo@bar.com",
        "firstname"   => "Joe",
        "lastname"    => "Bloggs",
        "name"        => "Joe Bloggs",
        "email"       => "foo@bar.com",
        "homePhone"   => "404 475 4840",
        "workPhone"   => "404 475 4840",
        "cellPhone"   => "404 475 4840",
        "homeAddress" => "Foo Lane, NY",
        "company"     => "Acme",
        "id"          => "uid=foo@bar.com",
        "__type"      => "Object",
        "source"      => "bar_clients"
      }

      horde.expects(:request).once.with('clients.getClient', id).returns response

      client = horde.get_client_by_id id
      client.should be response
    end
  end

  describe "searching for clients" do
    it "should return a set of client data" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      response = {'Acme' => [
        {
          "__key"       => "uid=foo@bar.com",
          "__uid"       => "foo@bar.com",
          "firstname"   => "Joe",
          "lastname"    => "Bloggs",
          "name"        => "Joe Bloggs",
          "email"       => "foo@bar.com",
          "homePhone"   => "404 475 4840",
          "workPhone"   => "404 475 4840",
          "cellPhone"   => "404 475 4840",
          "homeAddress" => "Foo Lane, NY",
          "company"     => "Acme",
          "id"          => "uid=foo@bar.com",
          "__type"      => "Object",
          "source"      => "bar_clients"
        }]
      }

      horde.expects(:request).once.with('clients.searchClients', ['Acme'], ['company']).returns response

      client = horde.search_clients :company => 'Acme'
      client.should be response
    end
  end

  describe "finding the first client by company name" do
    it "should return the relevant client data" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      client_data = {
        "__key"       => "uid=foo@bar.com",
        "__uid"       => "foo@bar.com",
        "firstname"   => "Joe",
        "lastname"    => "Bloggs",
        "name"        => "Joe Bloggs",
        "email"       => "foo@bar.com",
        "homePhone"   => "404 475 4840",
        "workPhone"   => "404 475 4840",
        "cellPhone"   => "404 475 4840",
        "homeAddress" => "Foo Lane, NY",
        "company"     => "Acme",
        "id"          => "uid=foo@bar.com",
        "__type"      => "Object",
        "source"      => "bar_clients"
      }
      response = {'Acme' => [client_data]}

      horde.expects(:request).once.with('clients.searchClients', ['Acme'], ['company']).returns response

      client = horde.first_client_for_company 'Acme'
      client.should be == client_data
    end

    context "if no matches are found" do
      it "should return nil" do
        horde = HordeRPC.new 'http://horde.bar.com/rpc/'

        response = {'Acme' => []}

        horde.expects(:request).once.with('clients.searchClients', ['Acme'], ['company']).returns response

        client = horde.first_client_for_company 'Acme'
        client.should be nil
      end
    end
  end

  describe "recording time" do
    it "should record time given the requested parameters" do
      horde = HordeRPC.new 'http://horde.bar.com/rpc/'

      options = {
        :date         => Date.today,
        :client       => 'bob@acme.com',
        :type         => 1,
        :hours        => 1.2,
        :description  => 'Did stuff',
        :employee     => 'foo@bar.com'
      }

      horde.expects(:request).once.with('time.recordTime', options)

      horde.record_time options
    end

    %w{date client type hours description employee}.each do |required_option|
      context "without specifying #{required_option}" do
        it "should raise ArgumentError" do
          horde = HordeRPC.new 'http://horde.bar.com/rpc/'

          options = {
            :date         => Date.today,
            :client       => 'bob@acme.com',
            :type         => 1,
            :hours        => 1.2,
            :description  => 'Did stuff',
            :employee     => 'foo@bar.com'
          }
          options.merge! required_option.to_sym => nil

          horde.expects(:request).never

          lambda { horde.record_time options }.should raise_error(ArgumentError)
        end
      end
    end
  end
end


lib/horde.rb
require 'horde/version'
require 'xmlrpc/client'

class Horde
  ##
  # Create a new instance of the Horde connection
  #
  # @param [String] uri XML-RPC URI for the horde instance
  # @param [String, Optional] username
  # @param [String, Optional] password
  #
  def initialize(uri, username = nil, password = nil)
    @xmlrpc_client = XMLRPC::Client.new2 uri
    @xmlrpc_client.user     = username
    @xmlrpc_client.password = password
  end

  ##
  # Make an RPC request against the Horde instance
  #
  # @param [String] method the method name to execute
  # @param [Object, Optional] options
  #
  # @return [Object] response to the RPC
  #
  def request(method, *options)
    @xmlrpc_client.call method, *options
  end

  ##
  # Return data for a client by its ID
  #
  # @param [String] id the clients ID
  #
  # @return [Hash,nil] client data
  #
  def get_client_by_id(id)
    request 'clients.getClient', id
  end

  ##
  # Search clients by KV-pairs
  #
  # @param [Hash<#to_s, String>] search_params
  #
  # @return [Hash] map of search terms to results
  #
  # @example Search for clients by first name
  #
  #   horde.search_clients :firstname => 'Bob'
  #   # => {'Acme' => [
  #          {
  #           "__key"       => "uid=foo@bar.com",
  #           "__uid"       => "foo@bar.com",
  #           "firstname"   => "Joe",
  #           "lastname"    => "Bloggs",
  #           "name"        => "Joe Bloggs",
  #           "email"       => "foo@bar.com",
  #           "homePhone"   => "404 475 4840",
  #           "workPhone"   => "404 475 4840",
  #           "cellPhone"   => "404 475 4840",
  #           "homeAddress" => "Foo Lane, NY",
  #           "company"     => "Acme",
  #           "id"          => "uid=foo@bar.com",
  #           "__type"      => "Object",
  #           "source"      => "bar_clients"
  #          }]
  #        }
  #
  def search_clients(search_params = {})
    # searchClients takes two arrays as params:
    # * list of search strings
    # * list of search fields
    request 'clients.searchClients', search_params.values, search_params.keys.map(&:to_s)
  end

  ##
  # Find the first client record for a particular company name
  #
  # @param [String] name the name of the Company
  #
  # @return [Hash,nil] data for the client associated with the requested company
  #
  def first_client_for_company(name)
    results = search_clients :company => name
    results[name].first
  end

  ##
  # Record time against a project
  #
  # @params [Hash] options
  # @option options [Date] :date the date of the billable activity
  # @option options [String] :client the ID of the client to bill
  # @option options [Integer] :type the type of billable activity
  # @option options [Float] :hours the number of hours to bill
  # @option options [String] :description a description of the billable activity
  # @option options [String] :employee the username of the employee performing the billable activity
  #
  def record_time(options)
    raise ArgumentError unless options.values_at(:date, :client, :type, :hours, :description, :employee).all?
    request 'time.recordTime', options
  end
end


lib/horde_rpc.rb
require 'horde_rpc/version'
require 'xmlrpc/client'

class HordeRPC
  ##
  # Create a new instance of the Horde connection
  #
  # @param [String] uri XML-RPC URI for the horde instance
  # @param [String, Optional] username
  # @param [String, Optional] password
  #
  def initialize(uri, username = nil, password = nil)
    @xmlrpc_client = XMLRPC::Client.new2 uri
    @xmlrpc_client.user     = username
    @xmlrpc_client.password = password
  end

  ##
  # Make an RPC request against the Horde instance
  #
  # @param [String] method the method name to execute
  # @param [Object, Optional] options
  #
  # @return [Object] response to the RPC
  #
  def request(method, *options)
    @xmlrpc_client.call method, *options
  end

  ##
  # Return data for a client by its ID
  #
  # @param [String] id the clients ID
  #
  # @return [Hash,nil] client data
  #
  def get_client_by_id(id)
    request 'clients.getClient', id
  end

  ##
  # Search clients by KV-pairs
  #
  # @param [Hash<#to_s, String>] search_params
  #
  # @return [Hash] map of search terms to results
  #
  # @example Search for clients by first name
  #
  #   horde.search_clients :firstname => 'Bob'
  #   # => {'Acme' => [
  #          {
  #           "__key"       => "uid=foo@bar.com",
  #           "__uid"       => "foo@bar.com",
  #           "firstname"   => "Joe",
  #           "lastname"    => "Bloggs",
  #           "name"        => "Joe Bloggs",
  #           "email"       => "foo@bar.com",
  #           "homePhone"   => "404 475 4840",
  #           "workPhone"   => "404 475 4840",
  #           "cellPhone"   => "404 475 4840",
  #           "homeAddress" => "Foo Lane, NY",
  #           "company"     => "Acme",
  #           "id"          => "uid=foo@bar.com",
  #           "__type"      => "Object",
  #           "source"      => "bar_clients"
  #          }]
  #        }
  #
  def search_clients(search_params = {})
    # searchClients takes two arrays as params:
    # * list of search strings
    # * list of search fields
    request 'clients.searchClients', search_params.values, search_params.keys.map(&:to_s)
  end

  ##
  # Find the first client record for a particular company name
  #
  # @param [String] name the name of the Company
  #
  # @return [Hash,nil] data for the client associated with the requested company
  #
  def first_client_for_company(name)
    results = search_clients :company => name
    results[name].first
  end

  ##
  # Record time against a project
  #
  # @params [Hash] options
  # @option options [Date] :date the date of the billable activity
  # @option options [String] :client the ID of the client to bill
  # @option options [Integer] :type the type of billable activity
  # @option options [Float] :hours the number of hours to bill
  # @option options [String] :description a description of the billable activity
  # @option options [String] :employee the username of the employee performing the billable activity
  #
  def record_time(options)
    raise ArgumentError unless options.values_at(:date, :client, :type, :hours, :description, :employee).all?
    request 'time.recordTime', options
  end
end
